### PR TITLE
Ch5.1

### PR DESCRIPTION
**Pull request by white-haired-uncle**
Time to pause for some review.

Mostly working on amplificators.  Each has its own theme, so the player will need to swap units/items/potions to deal with different strengths and weaknesses.  Castles are provided, since you can no longer just retreat and regroup it's only fair.
Amp1.  Very cold.  Enemies are good with cold, bad with fire.  Ice/snow expand each turn after 7.  Units that don't fly or move.frozen=1 who are on ice will break through, causing damage and poison (deep water).  I don't know if the amp caused this room to be cold or if he adapted to the cold.
2.  Fun with giving random enemies items (any weapons they can hold, plus an armor).  And there's evil versions of prophet and CM.  This amp is jealous of our heroes armies and is trying, with limited success, to build his own versions.  This needs to be made harder, but it is a lot of fun.
3.  There's lava.  Do the math.
4.   This amp likes to play with mechanical things.  So far, he's created suicide drones which use assassin ai to target E/L.  When they hit on offense, or are hit by impact, they explode doing self.hitpoints damage to themselves and all adjacent units.  I had three hit E earlier for like 150 damage in one turn.  It was awesome.  He had to turn tail and go heal.  Amp4 is working on drones that poison/incinerate and drones that do explosive slow, but he's not there yet.

So far, it's a lot more challenging.  Not hard, but no longer trivial.  And you need to use your army, not go forward with a few while the rest sit by the entrance/exit.

The "new" units are basically placeholders.  I have little interest and less talent in art and story. I know what I want their strengths and weaknesses to be and that's what I'm concentrating on.  Speaking of which, looks like I need to push those.

Need to export sort_unit_list() the right way.  No idea what that is at the moment.

When you control the undead in amp4 they were requiring upkeep.  A lot of upkeep.

Changed soulgate such that when you run out of units you only get very weak ones.  I don't see any reason to not just run out, but if you're going to get free recruits they should all be L0 (need some more units for this, like Elvish Civilians).

New graphical autorecall menu slipped in there as well.  Once I started using it I just couldn't go back.  Code needs cleanup, but otherwise it seems solid.

### CODE REVIEW

**Comment by white-haired-uncle** [comment]
I killed an amplificator with disintegrate and got a "no speaker" message.  I assumed it was because there was no second_unit set, but from what I've seen so far it looks like should work.  ++- 

While I was looking at the event, I noticed that it's not supposed to work on $bosses.  I'm wondering the amps should be in bosses.

Also, in main.lua:227 are we missing a couple of checks (unslowable, unpetrifyable)?

```
                        if not unit_to_harm.status.unpoisonable then
                                set_status("poisoned", _"poisoned", _"female^poisoned", "poison.ogg")
                        end
                        set_status("slowed", _"slowed", _"female^slowed", "slowed.wav")
                        set_status("petrified", _"petrified", _"female^petrified", "petrified.ogg")
```

File: scenarios5/20_Southern_Guardian_Room.cfg [hunk 1]
```diff
@@ -126,6 +205,14 @@
                     flat={UNREACHABLE}
                 [/movement_costs]
             [/effect]
+#ifdef HARD
+                                   [effect]
+                                        apply_to=new_ability
+                                        [abilities]
+                                            {ABILITY_INCREASE_RESISTANCE_AURA _"Song of Metal" 25}
```
Comment (Dugy) [inline scenarios5/20_Southern_Guardian_Room.cfg:212]: You wanted to change this, right?

File: scenarios5/20_Southern_Guardian_Room.cfg [hunk 2]
```diff
@@ -173,21 +268,41 @@
                 [movement_costs]
                     flat={UNREACHABLE}
                 [/movement_costs]
+#ifdef HARD
+                                   [effect]
+                                        apply_to=new_ability
+                                        [abilities]
+                                            {ABILITY_ALL_DAMAGE_AURA 30}
```
Comment (Dugy) [inline scenarios5/20_Southern_Guardian_Room.cfg:275]: Are you sure you wanted this one? This one increases also the damage of the user's enemies (that is, the players').

File: units/Amplificator.cfg [hunk 3]
```diff
@@ -32,19 +39,55 @@
...
+        fire=30
+        {QUANTITY fire 70 50 30}
+        cold=30
+        {QUANTITY code 70 50 30}
+        arcane=30
+        {QUANTITY arcane 70 50 30}
     [/resistance]
+    [attack]
+        name=claw
+        description= _"claw-like appendage"
```
Comment (Dugy) [inline units/Amplificator.cfg:57]: This won't be animated.

File: units/Fire_Reaper.cfg [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    id=Fire Reaper
+    name=_"Fire Reaper"
+    [base_unit]
+        id=Reaper
+    [/base_unit]
+    advances_to="Fire Reaper"
+    level=5
+    hide_help=true
+    do_not_list=yes
+    image="units/undead/reaper.png~RC(9d9da2,161515,141313>orange)~RC(b2b2b2,171616,151514>yellow)"
```
Comment (Dugy) [inline units/Fire_Reaper.cfg:12]: He'll change his colour back when an animation stats. You can prevent it by adding these colour changes as an `[effect]` that applies to `image_mod`.

File: units/Incendiary_Drone.cfg [hunk 5]
```diff
@@ -0,0 +1,119 @@
+#textdomain wesnoth-loti
+[unit_type]
+    id=Incendiary Drone
```
Comment (Dugy) [inline units/Incendiary_Drone.cfg:3]: Can you use `base_unit` here?

File: units/Phantom_of_the_Opera.cfg [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#textdomain wesnoth-loti
+[unit_type]
+    id=Phantom of the Opera
+    name= _"Phantom of the Opera"
+    advances_to=null
+        [base_unit]
```
Comment (Dugy) [inline units/Phantom_of_the_Opera.cfg:6]: Bad indentation.

File: units/Power_Generator.cfg [hunk 7]
```diff
@@ -24,7 +24,7 @@
     [/resistance]
     movement=0
     experience=150
-    level=3
+    level=8
```
Comment (Dugy) [inline units/Power_Generator.cfg:27]: Did you really intend to change a unit that only appears in chapter 9?

File: units/Watched_Pot_Drone.cfg [hunk 8]
```diff
@@ -0,0 +1,128 @@
+#textdomain wesnoth-loti
+[unit_type]
```
Comment (Dugy) [inline units/Watched_Pot_Drone.cfg:2]: Can you use `base_unit`?

File: utils/abilities.cfg [hunk 9]
```diff
@@ -965,6 +964,40 @@ Enemy units cannot see this unit during dawn or dusk, except if they have units
...
+            [/filter_location]
+        [/filter_self]
+        poison=cured
+    [/regenerate]
+#enddef
+#define ABILITY_PROPEL
+    [dummy]
+        id=propel
+        name= _ "propel"
+        description= _ "When an ally is standing next to this unit at the beginning of its turn, it gains two movement points."
```
Comment (Dugy) [inline utils/abilities.cfg:988]: There is an identical ability named `wardrums`. Please change it to 3 movement points or add some other difference. Or reuse `wardrums`.

File: utils/abilities.cfg [hunk 10]
```diff
@@ -1891,7 +1924,7 @@ Enemy units cannot see this unit during dawn or dusk, except if they have units
         name= _ "electrical discharge"
         female_name= _ "female^electrical discharge"
         description= _ "When this unit is hit with a melee attack, the resulting discharge of electricity causes an equal amount of lightning damage to be taken by the attacker. It can kill."
-#        halo_image_self="halo/reflect.png"  # need something that looks like it's throwing off sparks
+        #        halo_image_self="halo/reflect.png"  # need something that looks like it's throwing off sparks
```
Comment (Dugy) [inline utils/abilities.cfg:1927]: Please open an issue for this once this change is merged.

File: utils/abilities.cfg [hunk 11]
```diff
@@ -1962,81 +1995,81 @@ Enemy units cannot see this unit during dawn or dusk, except if they have units
...
+    [teleport]
+        id=recall
+        name= _ "recall"
+        female_name= _ "female^recall"
+        description= _ "This unit may teleport to an leader of its side."
+        [tunnel]
+            bidirectional=false
+            [source]
+                x=$teleport_unit.x,y=$teleport_unit.y
+            [/source]
```
Comment (Dugy) [inline utils/abilities.cfg:2007]: Did I commit some bad indentation here?

**Review by Dugy — CHANGES_REQUESTED** [review]

File: units/Powerslave.cfg [hunk 12]
```diff
@@ -0,0 +1,13 @@
+#textdomain wesnoth-loti
+[unit_type]
+    id=Powerslave
```
Comment (Dugy) [inline units/Powerslave.cfg:3]: I see you've figured out my naming patterns.

File: scenarios5/20_Southern_Guardian_Room.cfg (hunk 1 above)
Comment (white-haired-uncle) [reply scenarios5/20_Southern_Guardian_Room.cfg:212]: yes

File: scenarios5/15_Amplificator2.cfg [hunk 13]
```diff
@@ -231,6 +227,89 @@
...
+                    if item.sort == "armour" then  -- or item.sort == "armourword"?
+                        table.insert(armours,i)
+                    end
+                end
+                table.insert(items,loti.item.type[armours[mathx.random(#armours)]])
+
+                wml.array_access.set("my_items",items)
+            >>
+        [/lua]
+#ifdef NONONO
```
[the remark this answers is not in the document]
Comment (white-haired-uncle) [reply scenarios5/15_Amplificator2.cfg:280]: That is the me commenting out a block of code while I replaced it.   Turns out you can pass an array of tags to [insert_tag].

I have no idea how that got in there, as the only copies I have of that file have it removed.

File: scenarios5/20_Southern_Guardian_Room.cfg (hunk 2 above)
Comment (white-haired-uncle) [reply scenarios5/20_Southern_Guardian_Room.cfg:275]: No, don't want that.

That must be the new AMLA (elf sorc line?) that I never take.

**Review by white-haired-uncle — COMMENTED** [review]
didn't mean to hit the review button

File: units/Power_Generator.cfg (hunk 7 above)
Comment (white-haired-uncle) [reply units/Power_Generator.cfg:27]: They are much harder to kill now, so the player deserves more XP.
There's five of them in scenarios5/27_Power_Station.cfg

File: units/hybrids_of_steel.cfg [hunk 14]
```diff
@@ -71,7 +71,7 @@
         number=2
     [/attack]
 [/unit_type]
-[unit_type]
+[unit_type]  # A Spearman_steel advances directly to this (jumping level 4), but it's not much up an upgrade at all.
```
Comment (white-haired-uncle) [inline units/hybrids_of_steel.cfg:74]: I changed this in another branch.  Can I remove this file from the commit?

File: units/Phantom_of_the_Opera.cfg (hunk 6 above)
Comment (white-haired-uncle) [reply units/Phantom_of_the_Opera.cfg]: done.

File: units/Powerslave.cfg (hunk 12 above)
Comment (white-haired-uncle) [reply units/Powerslave.cfg:3]: I haven't noticed any other units named for old Iron Maiden songs (though Prophet_of_Disaster is a line from Die With Your Boots On, not a song title).  But then I gave up on them back in the early 90s when they, well I don't know what they did but it really sucked.

File: units/Incendiary_Drone.cfg (hunk 5 above)
Comment (white-haired-uncle) [reply units/Incendiary_Drone.cfg:3]: Probably.  I don't know what I'm doing here, so I probably just copied the file.

File: utils/abilities.cfg (hunk 9 above)
Comment (white-haired-uncle) [reply utils/abilities.cfg]: I knew I'd seen that somewhere.  Done.

File: utils/abilities.cfg (hunk 11 above)
Comment (white-haired-uncle) [reply utils/abilities.cfg:2000]: Could have been me, I've been doing a lot of cut and paste merging.

File: utils/abilities_events.cfg [hunk 15]
```diff
@@ -40,6 +40,65 @@
...
+                        id=$this_item.id
+                    [/filter]
+                    amount=${BOMBER}.hitpoints
+                    damage_type=${WEAPON_TYPE}
+                    fire_event=yes
+                    animate=yes
+                    kill=yes
+                [/harm_unit]
+            [/do]
+        [/foreach]
```
[the remark this answers is not in the document]
Comment (white-haired-uncle) [reply utils/abilities_events.cfg:99]: I'm not sure I follow the question, but I think I know the answer.

For purposes of animation, the drone needs to blow up (die) first.  So I have to store a list of his victims before he dies.

What's odd is that the victims take ${BOMBER}.hitpoints after he's dead.  I don't understand why that works (or where my changes went that stored that before he blew up).

One thing I don't like about the animation is each adjacent unit is damaged one at a time.  I'd like to find something like a non-blocking animation, so they could all appear to get hit at roughly the same time.

File: lua/recall.lua [hunk 16]
```diff
@@ -111,3 +111,425 @@ function wesnoth.wml_actions.recall_from_variable(cfg)
...
 	wml.variables[to_var] = picked - 1
 end
+-- luacheck: ignore 113
+-- luacheck: ignore 211
+-- luacheck: ignore 612
+
+local chat = wesnoth.interface.add_chat_message
+local sf = string.format
+
+function wesnoth.wml_actions.autorecall_menu()
```
[the remark this answers is not in the document]
Comment (white-haired-uncle) [reply lua/recall.lua:121]: This is the one I've been wanting for a long time.  One screen, no back and forth between add/remove menus, and the ability to adjust the order without having to remove/add units.
autorecall

[the remark this answers is not in the document]
Comment (Dugy) [reply lua/recall.lua:121]: Do you mean that you'll merge the two menus together later?

File: scenarios5/15_Amplificator2.cfg (hunk 13 above)
[the remark this answers is not in the document]
Comment (Dugy) [reply scenarios5/15_Amplificator2.cfg]: I suppose you plan to remove it.

File: units/Power_Generator.cfg (hunk 7 above)
Comment (Dugy) [reply units/Power_Generator.cfg:27]: Aha, okay.

File: utils/abilities_events.cfg (hunk 15 above)
[the remark this answers is not in the document]
Comment (Dugy) [reply utils/abilities_events.cfg:99]: There's no way to do that. What you can do is to make a big explosion animation whose effect seemingly hits all units, while the game engine would see only 1 hit. I can create that if you want.

File: units/hybrids_of_steel.cfg (hunk 14 above)
Comment (Dugy) [reply units/hybrids_of_steel.cfg:74]: Yes, but it's quite complicated. I think the easiest way to remove this is to erase that comment and include it in a commit in this feature branch.

File: units/Powerslave.cfg (hunk 12 above)
Comment (Dugy) [reply units/Powerslave.cfg:3]: I never quoted Iron Maiden specifically, but many names were inspired by metal songs (typically more obscure than Iron Maiden). Iron Maiden fits the scheme.

**Comment by Dugy** [comment]
> didn't mean to hit the review button

You didn't, but I haven't noticed that sorry.

File: scenarios5/15_Amplificator2.cfg (hunk 13 above)
[the remark this answers is not in the document]
Comment (white-haired-uncle) [reply scenarios5/15_Amplificator2.cfg]: Yes, I did, thanks.

File: lua/recall.lua (hunk 16 above)
[the remark this answers is not in the document]
Comment (white-haired-uncle) [reply lua/recall.lua:121]: Yes, this one will replace the one in both ch5 and ch9 after I finish with part1 and dig up and old save to test ch9.

File: units/Fire_Reaper.cfg (hunk 4 above)
Comment (white-haired-uncle) [reply units/Fire_Reaper.cfg]: Done, thanks.

File: lua/recall.lua (hunk 16 above)
[the remark this answers is not in the document]
Comment (Dugy) [reply lua/recall.lua:121]: All right.

File: utils/abilities_events.cfg (hunk 15 above)
[the remark this answers is not in the document]
Comment (Dugy) [reply utils/abilities_events.cfg:99]: Please open an issue for that once this is merged.

File: utils/abilities.cfg (hunk 10 above)
Comment (white-haired-uncle) [reply utils/abilities.cfg:1920]: noted

File: utils/abilities_events.cfg (hunk 15 above)
[the remark this answers is not in the document]
Comment (white-haired-uncle) [reply utils/abilities_events.cfg:99]: noted

File: units/Watched_Pot_Drone.cfg (hunk 8 above)
Comment (white-haired-uncle) [reply units/Watched_Pot_Drone.cfg:2]: I think I made a change to just about everything in the drones, so there's nothing really left to inherit from what I understand of base_unit.

They do need descriptions, and this one could probably use a name (I couldn't think of anything that sounds like a unit that makes things slow).

Comment (Dugy) [reply units/Watched_Pot_Drone.cfg:2]: The animation code is still duplicated.

**Comment by white-haired-uncle** [comment]
So, somewhere in there I'm pretty sure you pointed out the the Amplificator's claw attack (he needs a melee attack, it's so hard to get leaders to attack they should at least fight back) won't have an animation.  I can't find that, and while I see what you mean I don't know what to do about since there's no image I know of that would look good.

Other than that, I think we're good here.  There are a couple other things I can see in the long term:
1) Your idea about re-enabling upkeep after objectives are met.  This would need to be countered by more gold (my first thought is when you kill an enemy leader you take some percentage of his remaining gold).
2) Higher level recruits/spawns on HARD.

But those should probably wait until we see how the current changes play out.  I've played them quite a bit, and everything is winnable and not too difficult for me, but then I know exactly what to do.  I'm loathe to make even more fundamental changes unit there is more testing by someone else.

**Comment by Dugy** [comment]
> I can't find that, and while I see what you mean I don't know what to do about since there's no image I know of that would look good.

Now there is:

mutant-attack-1
mutant-attack-2

> Your idea about re-enabling upkeep after objectives are met. This would need to be countered by more gold (my first thought is when you kill an enemy leader you take some percentage of his remaining gold).

You can usually exit in the same turn as you complete the objectives, right? Or maybe 1 turn later. The random gold drops are not enough to cover this?

> Higher level recruits/spawns on HARD.

I think there is a risk a strong enough player would farm them for even more profit.

> But those should probably wait until we see how the current changes play out. I've played them quite a bit, and everything is winnable and not too difficult for me, but then I know exactly what to do.

That is also an option. I am leaving it up to you, I am fine with farming in this area since it's the last part of the campaign.

**Comment by white-haired-uncle** [comment]
> > I can't find that, and while I see what you mean I don't know what to do about since there's no image I know of that would look good.
> 
> Now there is:
> 
> mutant-attack-1 mutant-attack-2
> 
> > Your idea about re-enabling upkeep after objectives are met. This would need to be countered by more gold (my first thought is when you kill an enemy leader you take some percentage of his remaining gold).
> 
> You can usually exit in the same turn as you complete the objectives, right? Or maybe 1 turn later. The random gold drops are not enough to cover this?
> 
> > Higher level recruits/spawns on HARD.
> 
> I think there is a risk a strong enough player would farm them for even more profit.
> 
> > But those should probably wait until we see how the current changes play out. I've played them quite a bit, and everything is winnable and not too difficult for me, but then I know exactly what to do.
> 
> That is also an option. I am leaving it up to you, I am fine with farming in this area since it's the last part of the campaign.



> > I can't find that, and while I see what you mean I don't know what to do about since there's no image I know of that would look good.
> 
> Now there is:
> 

Very nice.  I may have gotten the order wrong.  I thought the top one was supposed to come first, and when I quoted your post I could see that it was mutant-attack-1 (when I go to save it it saves as aslkfjalksflfashf.png, not as the original name), but when he was attacking it didn't look right that way (my unit was to his NE, had he been to the south I think it would look better the other way).   But I even with no acceleration, it goes by so quickly I doubt it matters.

> 
> > Your idea about re-enabling upkeep after objectives are met. This would need to be countered by more gold (my first thought is when you kill an enemy leader you take some percentage of his remaining gold).
> 
> You can usually exit in the same turn as you complete the objectives, right? Or maybe 1 turn later. The random gold drops are not enough to cover this?

If you leave someone by the door, yes, unless you need a few turns to pick up dropped loot.  I'm also assuming that if you re-enter a room you've already cleared you'll be paying upkeep the whole time.

I've found that the only drawback to reducing the number of drops was less gold, to the point where I couldn't really grow my autorecall list, so I added a small drop for each "boss".

Of course, this gives value back to your loyal units, so maybe gold won't be as much of a factor as I suspect.

BTW, the way I remember it certain units (Ethiel is one, I think) used to be loyal but now aren't.  I'd like to see the "special" units (Mario, Ethiel, Connor, Skellie) loyal, as well as any you "save" (necromancer from Journey South, shadow prince(?) from Pirates who shows up too late to be much help, prisoners freed from Prison).  If not immediately, then at least when you get to the tunnels.

> 
> > Higher level recruits/spawns on HARD.
> 
> I think there is a risk a strong enough player would farm them for even more profit.

True.  Currently, I find the lower level hybrids to be irrelevant, and the higher level ones (not including the RG and GM who do that secret 10HP damage when hit by melee) to be just dangerous enough that I need to pay attention not to let them near injured units.  So I think higher level units on hard would be better.  But I don't think anything is really going to stop a good player who has taken care to maximize the strength of his army (which is perfectly fine IMO, if you know/work enough to build such an army you probably understand the consequences and have chosen to play that way).

> 
> > But those should probably wait until we see how the current changes play out. I've played them quite a bit, and everything is winnable and not too difficult for me, but then I know exactly what to do.
> 
> That is also an option. I am leaving it up to you, I am fine with farming in this area since it's the last part of the campaign.

That is a good point.  I usually skip straight to Akula's place, and skip chapter 9.5, because at that point building my army is irrelevant.  But there may be people who enjoy that part of the game, ruining it for them because I don't play that way wouldn't be cool.

**Comment by Dugy** [comment]
> it didn't look right that way (my unit was to his NE, had he been to the south I think it would look better the other way)

I have drawn it to fit a SE or S direction. I thought the file names would display.

> If you leave someone by the door, yes, unless you need a few turns to pick up dropped loot. I'm also assuming that if you re-enter a room you've already cleared you'll be paying upkeep the whole time.

It's possible to place an event that will trigger in 3 turns and bring back the upkeep. That should be long enough to collect stuff and leave. Some soldiers might voice this by complaining that he doesn't like to loiter when a dangerous enemy is finishing an evil plan.

> BTW, the way I remember it certain units (Ethiel is one, I think) used to be loyal but now aren't.

I gave them all an exclusive trait, because the loyal trait didn't do much to make the units unique.

I forgot about the Journey South necro and pirate leader.

> So I think higher level units on hard would be better. But I don't think anything is really going to stop a good player who has taken care to maximize the strength of his army (which is perfectly fine IMO, if you know/work enough to build such an army you probably understand the consequences and have chosen to play that way).

How about no units? They will stop coming because they are regrouping after the defeat.

**Comment by white-haired-uncle** [comment]
Made Akula's lightning attack actually lightning.  That always bugs me.  Adjusted the damage down to account for probable resistances.

Made the steel General/GM into much better leaders.  They are normally surrounded by units whose levels are high enough they wouldn't get any leadership without this.  The GM/RG REALLY need a halo and a warning about their secret 10HP melee return damage, probably Akula too.

The final battle is better, but still a ways to go.  It's very hard to make it hard.

**Comment by Dugy** [comment]
I suggest you not to spend too much on perfectionising it, or you risk losing motivation.

File: units/Watched_Pot_Drone.cfg (hunk 8 above)
Comment (white-haired-uncle) [reply units/Watched_Pot_Drone.cfg:2]: the sound is changed, so the only thing that would be inherited is a few lines of standing animation.

**Comment by white-haired-uncle** [comment]
Okay, that's it.  I played ch5 through again, once on hard, once on easy.  Easy still works correctly and is still very easy.

Hard is a whole lot harder.  Probably not enough, but enough for now.  It's still possible to make it to the final battle without losing any units, but you have to be extremely careful.  No unfair surprises or anything, but you need to pay attention to your units and the enemies strength and weaknesses, much like encountering packs of demons in part 2.  I don't think it's possible to take on Akula without losses, at least not for me.

**Review by Dugy — APPROVED** [review]

**Comment by Dugy** [comment]
Okay, merge it when you choose. You have some merge conflicts there, though.

**Pull request merged**